src/AzureAuth/Commands/Ado/CommandPat.cs

```diff
@@ -63,16 +90,101 @@ private enum OutputMode
         [Option(OutputOption, OutputHelp, CommandOptionType.SingleValue)]
         private OutputMode Output { get; set; } = OutputMode.Token;
 
+        [Option(DomainOption, DomainHelp, CommandOptionType.SingleValue)]
+        private string Domain { get; set; } = "microsoft.com";
+
         /// <summary>
         /// Executes the command and returns a status code indicating the success or failure of the execution.
         /// </summary>
         /// <param name="logger">The <see cref="ILogger{T}"/> instance that is used for logging.</param>
+        /// <param name="publicClientAuth">An <see cref="IPublicClientAuth"/>.</param>
         /// <param name="eventData">Lasso injected command event data.</param>
         /// <returns>An integer status code. 0 for success and non-zero for failure.</returns>
-        public int OnExecute(ILogger<CommandPat> logger, CommandExecuteEventData eventData)
+        public int OnExecute(ILogger<CommandPat> logger, IPublicClientAuth publicClientAuth, CommandExecuteEventData eventData)
         {
-            logger.LogInformation("coming soon");
+            var accessToken = this.AccessToken(publicClientAuth, eventData);
+            if (accessToken == null)
+            {
+                logger.LogError("Failed to acquire an Azure DevOps access token. Re-run with '--verbosity debug' for more info.");
+                return 1;
+            }
+
+            var patOptions = new PatOptions
+            {
+                Organization = this.Organization,
+                DisplayName = this.DisplayName,
+                Scopes = this.Scopes,
+            };
+
+            var cache = this.Cache();
+            var client = this.Client(accessToken.Token);
+            var manager = new PatManager(cache, client);
+
+            using (new CrossPlatLock(LockfilePath))
+            {
+                var pat = manager.GetPatAsync(patOptions).Result;
+
+                // Do not use logger to avoid printing PATs into log files.
+                Console.WriteLine(FormatPat(pat, this.Output));

```

Comment:
I wondered if you'd comment on that. 😜 Glad you did since it was kind of bugging me.

Here's why I didn't do what you suggest. `FormatPat` returns a `string` after formatting and keeps all I/O as the domain of the calling method. From where I sand, to not write the newline requires conditional logic either in the calling method (to interpret `null` or `string.Empty` as a reason to avoid calling `Console.WriteLine`) or moving the printing into `FormatPat` itself, making it impure.

I'm actually happy to do either and don't have any strong preferences. Is there an approach you prefer, @kyle-rader-msft?

Reply:
Personally, I think I lean towards giving up on functional purity, renaming `FormatPat` to `DisplayPat`, and just adding `Console.WriteLine`s to the `switch` expression within.

Reply:
Yep that's my recommendation. Use Write, and let the format add newlines as needed.